[16.0] stock_available_to_promise_release: manage backorders and allows unrelease

lmignon opened the pull request:
* [x] includes  #475 

By default, when an outgoing transfer is released, a backorder is created for
the remaining quantities for products not available at release time. This behaviour
can also be changed by stock route by setting the option "No backorder at release".
In such a case, moves created for unavailable products will remain in the original
outgoing picking and no backorder will be created. At the same time, outgoing
picking will not bet set to printed at release time if one move has been created
from a route whith the option "No backorder at release" set.

At the end of the picking process (when the picking is validated) of an outgoing
transfer, if a backorder is created, the chained moves can be automatically
unreleased. This behaviour is activated by setting the option "Unrelease on backorder"
on the stock picking type.

cc @jbaudoux

jbaudoux commented on stock_available_to_promise_release/models/stock_picking.py:
```diff
@@ -160,7 +160,13 @@ def _after_release_set_expected_date(self):
         )
-        self.scheduled_date = new_expected_date
+        move_to_update = []
+        for move in self.move_ids.filtered(lambda m: m.state == "assigned"):
```
all released move of the chain must have the new_expected_date

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -383,10 +385,10 @@ def _after_release_update_chain(self):
-            moves._action_assign()
-            moves = moves.mapped("move_orig_ids")
+        move_ids = self.ids
+        for ids in self._get_all_orig_ids().values():
```
Need to check it's not having side effect on stock dynamic routing

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -427,3 +429,54 @@ def _assign_picking_post_process(self, new=False):
+                "move_ids": tuple(self.ids),
+            },
+        )
+        res.update(dict(self.env.cr.fetchall()))
```
The chain of moves has just been created. Aren't all ids already in the cache? Do you have less queries with this change?

lmignon replied:
The chained moves are not into the cache in case of unrelease and when computing if an unreleased is allowed. The goal is to be sure that with just 1 additional SQL query we've all the information required to compute in batch the `unrelease_allowed`field regardless of the moves chain depth. The benefit may not be enough compared to the complexity. Furthermore, it is true that such a query requires an explicit flush of the cache for the stock.move model and so, not sure if this is a real benefit in the end. I'll change the code.

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -51,9 +52,50 @@ class StockMove(models.Model):
+            unrelease_allowed = (
+                user_is_allowed
+                and not move.need_release
+                and move.rule_id.available_to_promise_defer_pull
```
and picking_type_id.code == 'outgoing' like done here https://github.com/OCA/wms/pull/478/files#diff-05db432ffe85dc4ee9d20893829c7d8c96d484cd2cb12c0be7e0a70a949be3d9R29

lmignon replied:
:+1:

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -478,5 +542,35 @@ def _get_all_moves_from_chain(self, field):
+            if move_origs:
+                # avoid to propagate cancel to the original move
+                move_origs.write({"propagate_cancel": False})
+                move_origs._action_cancel()
```
Needs split before if `move_orig.product_qty` > `move.product_qty`.

Example use case - set a fixed procurement group on pick rule, pick moves are merged
* ship1 <- pack1 <- pickA
* ship2 <- pack2 <- pickA

Unreleasing ship1 needs to split the merged moves of pickA

lmignon replied:
seems to be far more complex.... we could have more than one move as origin and one origin could have more than one move as destination :thinking:

jbaudoux commented on stock_available_to_promise_release/models/stock_picking_type.py:
```diff
@@ -14,6 +14,12 @@ class StockPickingType(models.Model):
         "backorder are unreleased if they come from a route configured "
         "to manually release moves",
     )
+    set_printed_at_release = fields.Boolean(
```
This is required in case of backorder at release to ensure that 2 packs are not linked to a single ship.
Can you remind me why we would be interested by having this as an option?

lmignon replied:
we need it since the backorder creation is specified on the route and a picking is linked to many routes... we could also set printed at release only if no route is configured with 'no_backorder_at_release'.

lmignon committed:
[IMP] stock_available_to_promise_release: Makes the creation of backorder at release time configurable

lmignon committed:
[IMP] stock_available_to_promise_release: Add unrelease action on released moves

lmignon committed:
[IMP] stock_available_to_promise_release: Allows to unrelease move at backorder creation

lmignon committed:
[IMP] stock_available_to_promise_release: Explain available settings into documentation

lmignon commented:
@jbaudoux All comments implemented

sebalix commented:
@lmignon #475 has been merged

lmignon commented:
Thank you @sebalix! Rebase done.

lmignon committed:
[IMP] stock_available_to_promise_release: Makes the release process a specific processing step

* log the last release date
* add a configuration option to prevent the assignement of new moves to a released picking
* no more rely on the printed flag to control the backorder creation

lmignon committed:
[FIX] stock_available_to_promise_release: Fix stock.route form

The company_id field appears 2 times into the form. We must define a more precise xpath to point the mount point of our new fields into the original form

rousseldenis commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -49,9 +49,51 @@ class StockMove(models.Model):
+    def _compute_unrelease_allowed(self):
+        user_is_allowed = self.env.user.has_group("stock.group_stock_user")
+        for move in self:
+            unrelease_allowed = (
```
@lmignon Shouldn't this kind of rule in a separate method in order allow override?

lmignon replied:
hard to express this kind of check. I prefer to wait for the need of such a split...

rousseldenis commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +460,73 @@ def _assign_picking_post_process(self, new=False):
             self.picking_id.write({"priority": max(priorities)})
         return res
+
+    def _get_chained_moves_iterator(self, chain_field):
```
Shouldn't this method go in stock_helper module as it's quite generic?

rousseldenis commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +460,73 @@ def _assign_picking_post_process(self, new=False):
+        rounding = self.product_uom.rounding
+        new_origin_moves = self.env["stock.move"]
+        while float_compare(qty, 0, precision_rounding=rounding) > 0 and origins:
+            origin = origins[0]
```
I'd prefer using fields.first()

rousseldenis commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +460,73 @@ def _assign_picking_post_process(self, new=False):
+                new_origin_moves |= origin
+            else:
+                new_move_vals = origin._split(qty)
+                new_origin_moves |= self.create(new_move_vals)
```
What about using a set of ids to avoid recordset operations that can be bulky ?

lmignon replied:
because in this case, we don't expect a large recordset...

rousseldenis reviewed (commented):
LGTM. Some comments

lmignon committed:
[FIX] stock_available_to_promise_release: Exclude not releasable moves from release_ready computation

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +460,73 @@ def _assign_picking_post_process(self, new=False):
+            yield moves
+            moves = moves.mapped(chain_field)
+
+    def unrelease(self):
```
Add an action (wizard) to call this from the UI on the move AND on the picking

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +460,73 @@ def _assign_picking_post_process(self, new=False):
+                    lambda m: m.state not in ("done", "cancel")
+                )
+                if origin_moves:
+                    origin_moves = self._split_origins(origin_moves)
```
```suggestion
                    origin_moves = move._split_origins(origin_moves)
```

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -49,9 +49,51 @@ class StockMove(models.Model):
+
+    def _check_unrelease_allowed(self):
+        for move in self:
+            if not move.unrelease_allowed:
```
If one move of the picking chain is started, we should not be able to unrelease

lmignon commented:
@jbaudoux @rousseldenis Changes done

lmignon commented:
@jbaudoux @rousseldenis @sebalix Can we move forward with this one. I need it to be merged to avoid some trouble between all my pending PRs around the `stock_release_channel` addon.
For example in #493 I includes the last changes from @mt-software-de https://github.com/OCA/wms/pull/506/commits/bfb4f9b1ba5d8ae0607c9b4c71166587a6399cec but since this commit modify also the code into the `stock_available_to_promise_release`, It becomes impossible to use the current PR in my own project without making a glue branch for the 2 PRs... At the same time, it become impossible to tests #508 without this glue branch...

lmignon commented:
As temp fix I removed the forward port of https://github.com/OCA/wms/commit/bfb4f9b1ba5d8ae0607c9b4c71166587a6399cec from the migration of `stock_release_channel`

rousseldenis reviewed (approved):
LGTM. Code review

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +487,82 @@ def _assign_picking_post_process(self, new=False):
+            moves = moves.mapped(chain_field)
+
+    def unrelease(self, safe_unrelease=False):
+        """Unrelease unreleasavbe moves
```
typo

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +487,82 @@ def _assign_picking_post_process(self, new=False):
+            origins -= origin
+        return new_origin_moves
+
+    def _search_picking_for_assignation_domain(self):
```
With the refactor https://github.com/OCA/wms/pull/533/files#diff-fca2e9865c2d75a9793dbe699bffdf976606d996b69d09cacd25660f2926e998R410 you don't need to hack this method anymore. There is a new method for moving the moves to a backorder that can be improved with the new flag

lmignon replied:
@jbaudoux I would prefer to finish this and the following PRs before forward porting the latest changes from V14 as it becomes too complicated and creates too many conflicts to integrate new changes from 14 into the current PRs.

jbaudoux replied:
I see if we can merge first in v14 the backport of the new features you added in this migration. Then I'll rebase the refactor and it will be a lot easier to fwd port to this branch.

TDu committed:
stock_available_to_promise: Add 2 steps delivery test

TDu committed:
stock_available_to_promise: unrelease chain on cancel move

When a move is being cancelled if possible cancel all the related moves.
This will also be triggered when cancelling a sales order.

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +487,82 @@ def _assign_picking_post_process(self, new=False):
+                    impacted_picking_ids.update(origin_moves.mapped("picking_id").ids)
+                    # avoid to propagate cancel to the original move
+                    origin_moves.write({"propagate_cancel": False})
+                    origin_moves._action_cancel()
```
You have to cancel starting of the first origin of the chain and go down the chain. Easier it to make a complete recordset and cancel them all together.
So in case of pick/pack/ship, you cannot cancel pack then pick. You have to cancel pick, then pack, or pick+pack together.

jbaudoux commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -419,3 +487,82 @@ def _assign_picking_post_process(self, new=False):
+                qty -= origin.product_qty
+                new_origin_moves |= origin
+            else:
+                new_move_vals = origin._split(qty)
```
move_dest_ids on new_move_vals must be changed to be linked only on the next move that is concerned.
On the origin move, the next concerned move must be removed from move_dest_ids.

jbaudoux replied:
Note: This remark above is to be compatible with the module stock_picking_restrict_cancel_with_orig_move which ensures you don't cancel a move in the middle of a chain of moves (because it's bad)

jbaudoux reviewed (changes requested):
I did some testing and had issues to unrelease. A fix proposal is on its way.

TDu committed:
stock_available_to_promise: add 3 steps delivery test

jbaudoux commented on stock_available_to_promise_release/views/stock_picking_type_views.xml:
```diff
@@ -25,4 +25,22 @@
     </record>
+
+    <record id="stock_picking_type_form" model="ir.ui.view">
+        <field name="name">stock.picking.type.kanban</field>
```
typo in the name. It's not the kanban

jbaudoux commented on stock_available_to_promise_release/models/stock_picking_type.py:
```diff
@@ -8,6 +8,17 @@ class StockPickingType(models.Model):
+    prevent_new_move_after_release = fields.Boolean(
+        string="Prevent new move after release",
+        help="If checked, when a picking is released it's no more possible to "
+        "assign new moves to it.",
```
```suggestion
        help="If checked, once a delivery picking has been released, no more moves will be added to it. For example, if you add lines in the origin sales order, the new moves will be added to a new delivery.",
```

jbaudoux commented on stock_available_to_promise_release/models/stock_picking_type.py:
```diff
@@ -8,6 +8,17 @@ class StockPickingType(models.Model):
     count_picking_need_release = fields.Integer(compute="_compute_picking_count")
+    unrelease_on_backorder = fields.Boolean(
+        string="Unrelease on backorder",
+        help="If checked, when a backorder is created the moves into the "
+        "backorder are unreleased if they come from a route configured "
+        "to manually release moves",
```
```suggestion
        help="If checked, when a backorder is created (i.e. at the validation of the delivery), the moves into the "
        "backorder are unreleased as long as they came from a route configured "
        "to manually release moves. This means that the moves that were not delivered are put back in need for release and the unprocessed internal pulled moves are canceled.",
```

jbaudoux commented on stock_available_to_promise_release/models/stock_picking.py:
```diff
@@ -171,3 +179,20 @@ def action_open_move_need_release(self):
+        If safe_unrelease is True, the unreleasaable moves for which the
+        processing has already started will be ignored
+        """
+        self.mapped("move_ids").unrelease(safe_unrelease=safe_unrelease)
```
no need for mapped

jbaudoux commented on stock_available_to_promise_release/models/stock_picking.py:
```diff
@@ -171,3 +179,20 @@ def action_open_move_need_release(self):
+    def unrelease(self, safe_unrelease=False):
+        """Unrelease the moves of the picking.
+
+        If safe_unrelease is True, the unreleasaable moves for which the
```
typo
```suggestion
        If safe_unrelease is True, the unreleasable moves for which the
```

jbaudoux commented:
@lmignon any reason why you always wrote "Un Release" instead of "Unrelease" ? (cc @TDu )

lmignon commented:
> @lmignon any reason why you always wrote "Un Release" instead of "Unrelease" ? (cc @TDu )

@jbaudoux my frenchglish...

lmignon commented:
@jbaudoux The management of the backorders introduce some side-effect into the computation of counters on the view.
With this last change, when some products are not available at release time we no more put the move into a new backorder pick (its done when the channel is put to sleep). Unfortunately some of the counters and fields computed from domains provided by the `_field_picking_domains` method (https://github.com/OCA/wms/pull/493/files#diff-a8bb5a8dfb388e49242b1d9fe2274b33dccde4e032d0983647f70fa52399ecbcR174) rely on the predicate that work is to be done as soon as a ship no more need to be released (the concept of partially released ship doesn't exist)

jbaudoux commented:
> I did some testing and had issues to unrelease. A fix proposal is on its way.

The fixes have been done in the backport on v14. They are in those commits that needs to be integrated here:

- https://github.com/OCA/wms/pull/556/commits/ea271a69769cd8fdab244562fdb955e3bfafc8b8
- https://github.com/OCA/wms/pull/556/commits/d3b09534492964a7ab6f22f6ca564b11d9a7b2e1
- https://github.com/OCA/wms/pull/556/commits/773f402d3c23d9a1f483a45613065b1f162060db

lmignon committed:
[FIX] stock_available_to_promise_release: Typos

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

lmignon commented:
@jbaudoux commits included

lmignon committed:
[FIX] stock_available_to_promise: Fixes after forward port

Enhancements from 14.0 are provided by the previous 3 commits. This commit fixes the code to be compliant with 16.0

lmignon commented:
@jbaudoux @rousseldenis This one is ready to be merged...

lmignon commented:
@OCA/logistics-maintainers Someone to merge this one?

jbaudoux commented on stock_available_to_promise_release/migrations/16.0.1.0.0/post-migrate.py:
```diff
@@ -0,0 +1,14 @@
+        "stock_available_to_promise_release: Configure picking types"
+        " to prevent new moves into released pickings"
+    )
+    cr.execute("update stock_picking_type set prevent_new_move_after_release = true")
```
This should be in a 16.0.2.0.0 migration script as this is also required to run on instances having already the module running in 16.0

jbaudoux reviewed (commented):
Can you change the version to 16.0.2.0.0 and make the migration script for that version?

lmignon commented:
@jbaudoux Changes done and commit history cleaned up

jbaudoux commented:
/ocabot merge nobump
